esp-hal/src/sdio.rs

```diff
@@ -0,0 +1,208 @@
+    tuning::{TuningMode, TuningWidth},
+};
+
+mod slc;
+mod slchost;
+
+pub use slc::*;
+pub use slchost::*;
+
+/// SDIO peripheral instance.
+pub trait PeripheralInstance: crate::private::Sealed {
+    /// Represents the peripheral information type containing the register
+    /// block.
+    type Info;
+
+    /// Gets a static shared reference to the peripheral information.
+    fn info(&self) -> &'static Self::Info;
+}
+
+/// Represents the SDIO 2.0 peripheral for the microcontroller.
+#[derive(Debug)]
+pub struct Sdio<'d> {
+    slc: AnySlc<'d>,
+    slchost: AnySlchost<'d>,
+}
+
+impl<'d> Sdio<'d> {
+    /// Creates a new [Sdio].
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// use esp_hal::sdio::Sdio;
+    ///
+    /// use crate::peripheral::Peripherals;
+    ///
+    /// let dp = Peripheral::take().unwrap();
+    /// let _sdio = Sdio::new(dp.slc, dp.slchost);
+    /// ```
+    pub fn new(slc: impl SlcInstance + 'd, slchost: impl SlchostInstance + 'd) -> Self {
+        Self {
+            slc: slc.degrade(),
+            slchost: slchost.degrade(),
+        }
+    }
+
+    /// Gets a static reference to the SLC information.
+    pub fn slc(&self) -> &'static SlcInfo {
+        self.slc.info()
+    }
+
+    /// Gets a static reference to the SLCHOST information.
+    pub fn slchost(&self) -> &'static SlchostInfo {
+        self.slchost.info()
+    }
+}
+
+/// Represents the error variants for SDIO peripherals.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum Error {
+    /// Indicates a general error occured.
+    General,
+    /// Indicates use of an illegal command.
+    IllegalCommand,
+    /// Indicates a CRC error from the previous command.
+    Crc,
+    /// The function and/or type is unimplemented.
+    Unimplemented,
+}
+
+impl Error {
+    /// Creates a new [Error].
+    pub const fn new() -> Self {
+        Self::Unimplemented
+    }
+
+    /// Creates an general [Error].
+    #[inline]
+    pub const fn general() -> Self {
+        Self::General
+    }
+
+    /// Creates an illegal command [Error].
+    #[inline]
+    pub const fn illegal_command() -> Self {
+        Self::IllegalCommand
+    }
+
+    /// Creates an crc [Error].
+    #[inline]
+    pub const fn crc() -> Self {
+        Self::Crc
+    }
+
+    /// Creates an unimplemented [Error].
+    #[inline]
+    pub const fn unimplemented() -> Self {
+        Self::Unimplemented
+    }
+}
+
+impl Default for Error {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl core::fmt::Display for Error {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            Self::General => write!(f, "general"),
+            Self::IllegalCommand => write!(f, "illegal command"),
+            Self::Crc => write!(f, "CRC"),
+            Self::Unimplemented => write!(f, "unimplemented"),
+        }
+    }
+}
+
+impl core::error::Error for Error {}
+
+impl<'d> MmcOps for Sdio<'d> {
```

Comment:
Something is not quite right here. `MmcOps` is meant for the master side of this protocol, but you've implemented this on the slave peripheral.

I think you want the SD/MMC peripheral (#1928) and not the SDIO peripheral, right?

Reply:
> Something is not quite right here. `MmcOps` is meant for the master side of this protocol, but you've implemented this on the slave peripheral.
> 
> I think you want the SD/MMC peripheral (#1928) and not the SDIO peripheral, right?

Correct, the `MmcOps` trait is currently for the host, and I need to split the trait into potentially three traits:

- `MmcCommon`
- `MmcHost`
- `MmcDevice`

I'll keep working on the e-h trait along with the work in this pull request. Thanks for the review.